src/serac/physics/solid_functional.hpp

```diff
@@ -525,59 +484,138 @@ class SolidFunctional : public BasePhysics {
-            functional_call_args_[0] = d2u_dt2;
+    time_ += dt;
 
-            auto M = serac::get<1>((*M_functional_)(functional_call_args_, Index<0>{}));
-            std::unique_ptr<mfem::HypreParMatrix> m_mat(assemble(M));
+    // the 30 lines of code below are essentially equivalent to the 1-liner
+    // u += dot(inv(J), dot(J_elim[:, dofs], (U(t + dt) - u)[dofs]));
+    {
+      for (auto& bc : bcs_.essentials()) {
+        bc.projectBdrToDofs(du_, time_);
+      }
 
-            // J = M + c0 * H(u_predicted)
-            mfem::Vector K_arg(u_.Size());
-            add(1.0, u_, c0_, d2u_dt2, K_arg);
-            functional_call_args_[0] = K_arg;
+      auto & constrained_dofs = bcs_.allEssentialTrueDofs();
+      for (int i = 0; i < constrained_dofs.Size(); i++) {
+        du_[constrained_dofs[i]] -= displacement_(constrained_dofs[i]);
+      }
 
-            auto K = serac::get<1>((*K_functional_)(functional_call_args_, Index<0>{}));
+      dr_ = 0.0;
+      for (const auto& bc : bcs_.essentials()) {
+        bc.eliminateToRHS(*J_, du_, dr_);
+      }
 
-            functional_call_args_[0] = u_;
+      auto& lin_solver = nonlin_solver_.LinearSolver();
 
-            std::unique_ptr<mfem::HypreParMatrix> k_mat(assemble(K));
+      lin_solver.SetOperator(*J_);
 
-            J_.reset(mfem::Add(1.0, *m_mat, c0_, *k_mat));
-            bcs_.eliminateAllEssentialDofsFromMatrix(*J_);
+      lin_solver.Mult(dr_, du_);
 
-            return *J_;
-          });
+      displacement_ += du_;
+
+      for (auto& bc : bcs_.essentials()) {
+        bc.projectBdrToDofs(du_, time_);
+      }
+
+      // do I have to do this?
+      nonlin_solver_.SetOperator(*residual_with_bcs_);
+    }
+
+    nonlin_solver_.Mult(zero_, displacement_);
+  }
+
+  /**
+   * @brief Advance the timestep
+   *
+   * @param[inout] dt The timestep to attempt. This will return the actual timestep for adaptive timestepping
+   * schemes
+   * @pre SolidFunctional::completeSetup() must be called prior to this call
+   */
+  void advanceTimestep(double& dt) override
+  {
+    SLIC_ERROR_ROOT_IF(!residual_, "completeSetup() must be called prior to advanceTimestep(dt) in SolidFunctional.");
+
+    // Set the mesh nodes to the reference configuration
+    if (geom_nonlin_ == GeometricNonlinearities::On) {
+      mesh_.NewNodes(*reference_nodes_);
+    }
+
+    bcs_.setTime(time_);
+
+    if (is_quasistatic_) {
+      quasiStaticSolve(dt);
+    } else {
+      ode2_.Step(displacement_, velocity_, time_, dt);
+    }
+
+    // if our simulation involves materials that have internal state 
+    if (residual_->material_state_buffers_.size()) {
+
+      // after finding displacements that satisfy equilibrium, 
+      // compute the residual one more time, this time enabling
+      // the material state buffers to be updated
+      residual_->update_qdata = true;
+      (*residual_)(displacement_, zero_, parameter_states_[parameter_indices] ...);
```

Comment:
Could we take advantage of this additional call to take care of other business?

- If adjoint sensitivities are requested from material parameters, it would be efficient to promote them to dual numbers in the material operator at this call site only.  This will avoid unnecessary operations during the equilibrium search. (Furthermore, materials with iterative procedures can detect when the parameters are dual and do the adjoint computation by hand at the solution, rather than accumulating through the iteration).
- If we can turn off the zeroing out of the residual terms corresponding to essential BCs, then those slots will collect the reactions, for use in output and QoI.

Reply:
> If we can turn off the zeroing out of the residual terms corresponding to essential BCs, then those slots will collect the reactions, for use in output and QoI.

👍 

`Functional` doesn't know anything about essential b.c.s, so the output of this call will have the nodal reaction forces. Instead of discarding the output from the highlighted code, we're storing it now and there's a getter `SolidFunctional::nodalForces()` to access it.

> If adjoint sensitivities are requested from material parameters, it would be efficient to promote them to dual numbers in the material operator at this call site only. This will avoid unnecessary operations during the equilibrium search. (Furthermore, materials with iterative procedures can detect when the parameters are dual and do the adjoint computation by hand at the solution, rather than accumulating through the iteration).

For reference, the derivatives w.r.t. parameter fields aren't calculated during the equilibrium solve, they're lazily-evaluated when the user calls "computeSensitivity()`.